setup.py

```diff
@@ -98,23 +111,77 @@ def get_geos_paths():
 
 
 # Add numpy include dirs without importing numpy on module level.
-# See https://stackoverflow.com/questions/19919905/
-# how-to-bootstrap-numpy-installation-in-setup-py/21621689#21621689
+# derived from scikit-hep:
+# https://github.com/scikit-hep/root_numpy/pull/292
 class build_ext(_build_ext):
     def finalize_options(self):
         _build_ext.finalize_options(self)
         # Prevent numpy from thinking it is still in its setup process:
-        __builtins__.__NUMPY_SETUP__ = False
+        try:
+            del builtins.__NUMPY_SETUP__
+        except AttributeError:
+            pass
+
         import numpy
 
         self.include_dirs.append(numpy.get_include())
 
 
-module_lib = Extension(
-    "pygeos.lib",
-    sources=["src/lib.c", "src/geos.c", "src/pygeom.c", "src/ufuncs.c", "src/coords.c", "src/strtree.c"],
-    **get_geos_paths()
-)
+ext_modules = []
+
+if "clean" not in sys.argv:
+    ext_options = get_geos_paths()
+
+    ext_modules = [
+        Extension(
+            "pygeos.lib.core",
+            sources=[
+                "src/core.c",
+                "src/c_api.c",
+                "src/coords.c",
+                "src/geos.c",
+                "src/pygeom.c",
+                "src/strtree.c",
+                "src/ufuncs.c",
+            ],
+            **ext_options,
+        )
+    ]
+
+    if os.path.exists("MANIFEST.in"):
```

Comment:
The trick of looking for `MANIFEST.in` was inspired by [`fiona`](https://github.com/Toblerity/Fiona/blob/master/setup.py).  Basically, this is excluded from distributed packages, so if we see it, there is a high likelihood it is the source code for this repo, and needs to be built.

This could probably be moved to the top level above compiling the C extension.

Reply:
I'm not entirely sure I agree with the Cython recommendation here, but my concerns primarily are related to adding the generated C files to the repository:
* Cython is very easy to install compared to ensuring an adequate C build system with headers required here; i.e., it should be a trivial install on most platforms.
* the C files it generates are BIG and will likely lead to big changesets for small changes in Cython files; these will be a distraction during review.  But - there might be a way to include them in source distributions without including them in the repository (still need to work out packaging details here)
* the other Python projects that leverage Cython I used as reference here ([fiona .gitignore](https://github.com/Toblerity/Fiona/blob/master/.gitignore), [rasterio .gitignore](https://github.com/mapbox/rasterio/blob/master/.gitignore)) do not include the generated C files in their repositories.

Let me do a bit more research on the packaging side and come back to this.

Reply:
Ah, yes.  Fiona specifies the C files as a fallback when not building from a repository.  We should do the same here, which is in line with Cython recommendations.  Apologies for my confusion here.

Reply:
When building a sdist `python setup.py sdist`, does pygeos do the `cythonize` and does it include the `.c` files into the sdist? That's worth a test.

Reply:
Yes, `sdist` Cythonizes the Cython extensions and outputs C files.  I had to add a bit more to `MANIFEST.in` to make sure these were included properly in the distribution.

Reply:
On shipping the generated C files: projects like numpy and pandas stopped doing this (see https://github.com/numpy/numpy/pull/14453 for their rationale)

Reply:
If numpy and pandas stopped doing this, then I am :+1: on also not distributing them